File: backend/feedback_models.py
from pydantic import BaseModel, Field
from typing import Optional, List
from sqlmodel import SQLModel, Field as SQLField, Column, ForeignKey
from datetime import datetime
from sqlalchemy.dialects.postgresql import TEXT, ARRAY

# ====================================================================
# SQLModel table
# ====================================================================
class Feedbacks(SQLModel, table=True):
    __tablename__ = "feedbacks"
    id: Optional[int] = SQLField(default=None, primary_key=True)
    send_id: int = SQLField(foreign_key="profiles.id")
    rev_id: int = SQLField(foreign_key="profiles.id")
    group_id: Optional[int] = SQLField(default=None, foreign_key="travel_groups.id")
    group_image_url: Optional[str] = SQLField(default=None, sa_column=Column(TEXT))
    rating: Optional[int] = SQLField(default=None, sa_column=Column("rating", nullable=True))
    content: Optional[List[str]] = SQLField(default=None, sa_column=Column(ARRAY(TEXT)))
    anonymous: bool = SQLField(default=False)
    created_at: Optional[datetime] = SQLField(default=None, sa_column_kwargs={"default": "NOW()"})

# ====================================================================
# Pydantic models for API input/output
# ====================================================================
class CreateFeedbackInput(BaseModel):
    rev_id: int = Field(..., description="ID profile người nhận")
    group_id: int = Field(..., description="ID của nhóm du lịch (bắt buộc)")
    group_image_url: Optional[str] = Field(None, description="URL ảnh đại diện của group")
    rating: Optional[int] = Field(None, ge=1, le=5)
    content: Optional[List[str]] = Field(None, description="List tags cho nội dung đánh giá (e.g., ['friendly', 'punctual'])")
    anonymous: Optional[bool] = Field(False)

class UpdateFeedbackInput(BaseModel):
    rating: Optional[int] = Field(None, ge=1, le=5)
    content: Optional[List[str]] = Field(None, description="List tags cho nội dung đánh giá")
    anonymous: Optional[bool] = None

class FeedbackPublic(BaseModel):
    id: int
    send_id: int
    rev_id: int
    group_id: Optional[int]
    group_image_url: Optional[str]
    rating: Optional[int]
    content: Optional[List[str]]
    anonymous: bool
    created_at: Optional[datetime]

    class Config:
        from_attributes = True

class FeedbackDetail(FeedbackPublic):
    sender_name: Optional[str]  # Chỉ hiển thị nếu không anonymous
    sender_email: Optional[str]  # Email sender (ẩn nếu anonymous)
    receiver_name: Optional[str]  # Tên người nhận
    receiver_email: Optional[str]  # Email người nhận

# ====================================================================
# Models cho My Reputation
# ====================================================================
class GroupReputationSummary(BaseModel):
    group_id: int
    group_name: str
    group_image_url: Optional[str]
    feedbacks: List[FeedbackDetail]

class MyReputationResponse(BaseModel):
    average_rating: float
    total_feedbacks: int
    groups: List[GroupReputationSummary]

# ====================================================================
# Models cho Pending Reviews
# ====================================================================
class UnreviewedMember(BaseModel):
    profile_id: int
    profile_uuid: str
    email: str
    fullname: Optional[str]

class PendingReviewGroup(BaseModel):
    group_id: int
    group_name: str
    group_image_url: Optional[str]
    unreviewed_members: List[UnreviewedMember]

class PendingReviewsResponse(BaseModel):
    pending_groups: List[PendingReviewGroup]

File: frontend/lib/models/feedback_models.dart
<<<<<<< HEAD
=======
// Hàm tiện ích giúp chuyển đổi số an toàn (tránh lỗi crash khi server trả String)
double _parseDouble(dynamic value) {
  if (value == null) return 0.0;
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value) ?? 0.0;
  return 0.0;
}

>>>>>>> week10
class UnreviewedMember {
  final int profileId;
  final String fullname;
  final String email;

  UnreviewedMember({
    required this.profileId,
    required this.fullname,
    required this.email,
  });

  factory UnreviewedMember.fromJson(Map<String, dynamic> json) {
    return UnreviewedMember(
<<<<<<< HEAD
      profileId: json['profile_id'],
      // Fallback nếu tên null thì hiển thị default
=======
      profileId: json['profile_id'] is int ? json['profile_id'] : int.tryParse(json['profile_id'].toString()) ?? 0,
>>>>>>> week10
      fullname: json['fullname'] ?? 'Thành viên',
      email: json['email'] ?? '',
    );
  }
}

class PendingReviewGroup {
  final int groupId;
  final String groupName;
  final String? groupImageUrl;
  final List<UnreviewedMember> unreviewedMembers;

  PendingReviewGroup({
    required this.groupId,
    required this.groupName,
    this.groupImageUrl,
    required this.unreviewedMembers,
  });

  factory PendingReviewGroup.fromJson(Map<String, dynamic> json) {
    var list = json['unreviewed_members'] as List? ?? [];
    List<UnreviewedMember> membersList = list.map((i) => UnreviewedMember.fromJson(i)).toList();

    return PendingReviewGroup(
<<<<<<< HEAD
      groupId: json['group_id'],
=======
      groupId: json['group_id'] is int ? json['group_id'] : int.tryParse(json['group_id'].toString()) ?? 0,
>>>>>>> week10
      groupName: json['group_name'] ?? 'Nhóm du lịch',
      groupImageUrl: json['group_image_url'],
      unreviewedMembers: membersList,
    );
  }
}

<<<<<<< HEAD
// --- CÁC MODEL MỚI CHO REPUTATION ---
=======
// --- CÁC MODEL CHO REPUTATION (Đã sửa lỗi parse số) ---
>>>>>>> week10

class FeedbackDetail {
  final int id;
  final double rating;
  final List<String> content; // Tags
  final String? senderName;
  final bool anonymous;

  FeedbackDetail({
    required this.id,
    required this.rating,
    required this.content,
    this.senderName,
    required this.anonymous,
  });

  factory FeedbackDetail.fromJson(Map<String, dynamic> json) {
    return FeedbackDetail(
<<<<<<< HEAD
      id: json['id'],
      rating: (json['rating'] ?? 0).toDouble(),
=======
      id: json['id'] is int ? json['id'] : int.tryParse(json['id'].toString()) ?? 0,
      // SỬA: Dùng hàm _parseDouble an toàn
      rating: _parseDouble(json['rating']),
>>>>>>> week10
      content: (json['content'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
      senderName: json['sender_name'],
      anonymous: json['anonymous'] ?? false,
    );
  }
}

class GroupReputationSummary {
  final int groupId;
  final String groupName;
  final String? groupImageUrl;
  final List<FeedbackDetail> feedbacks;

  GroupReputationSummary({
    required this.groupId,
    required this.groupName,
    this.groupImageUrl,
    required this.feedbacks,
  });

  factory GroupReputationSummary.fromJson(Map<String, dynamic> json) {
    var list = json['feedbacks'] as List? ?? [];
    List<FeedbackDetail> fbList = list.map((i) => FeedbackDetail.fromJson(i)).toList();

    return GroupReputationSummary(
<<<<<<< HEAD
      groupId: json['group_id'],
=======
      groupId: json['group_id'] is int ? json['group_id'] : int.tryParse(json['group_id'].toString()) ?? 0,
>>>>>>> week10
      groupName: json['group_name'] ?? 'Nhóm',
      groupImageUrl: json['group_image_url'],
      feedbacks: fbList,
    );
  }
}

class MyReputationResponse {
  final double averageRating;
  final int totalFeedbacks;
  final List<GroupReputationSummary> groups;

  MyReputationResponse({
    required this.averageRating,
    required this.totalFeedbacks,
    required this.groups,
  });

  factory MyReputationResponse.fromJson(Map<String, dynamic> json) {
    var list = json['groups'] as List? ?? [];
    List<GroupReputationSummary> groupList = list.map((i) => GroupReputationSummary.fromJson(i)).toList();

    return MyReputationResponse(
<<<<<<< HEAD
      averageRating: (json['average_rating'] ?? 0).toDouble(),
      totalFeedbacks: json['total_feedbacks'] ?? 0,
=======
      // SỬA: Dùng hàm _parseDouble an toàn
      averageRating: _parseDouble(json['average_rating']),
      totalFeedbacks: json['total_feedbacks'] is int ? json['total_feedbacks'] : int.tryParse(json['total_feedbacks'].toString()) ?? 0,
>>>>>>> week10
      groups: groupList,
    );
  }
}